python-for-android/dists/kolibri/src/main/java/org/kivy/android/PythonWorker.java

```diff
@@ -112,37 +117,31 @@ public ListenableFuture<Result> startRemoteWork() {
         // This is somewhat similar to what the plain `Worker` class does, except that we
         // use `submit` instead of `execute` so we can propagate cancellation
         // See https://android.googlesource.com/platform/frameworks/support/+/60ae0eec2a32396c22ad92502cde952c80d514a0/work/workmanager/src/main/java/androidx/work/Worker.java
-        RunnableFuture<?> threadFuture = (RunnableFuture<?>)executor.submit(new Runnable() {
-            @Override
-            public void run() {
-                try {
-                    Result r = doWork();
-                    future.set(r);
-                }  catch (Exception e) {
-                    if (getRunAttemptCount() > MAX_WORKER_RETRIES) {
-                        Log.e(TAG, id + " Exception in remote python work", e);
-                        future.setException(e);
-                    } else {
-                        Log.w(TAG, id + " Exception in remote python work, scheduling retry", e);
-                        future.set(Result.retry());
-                    }
-                } finally {
-                    cleanup();
+        Future<?> threadFuture = executor.submit(() -> {
+            try {
+                Result r = doWork();
+                future.set(r);
+            }  catch (Exception e) {
+                if (!DO_RETRY || getRunAttemptCount() > MAX_WORKER_RETRIES) {
+                    Log.e(TAG, id + " Exception in remote python work", e);
+                    future.setException(e);
+                } else {
+                    Log.w(TAG, id + " Exception in remote python work, scheduling retry", e);
+                    future.set(Result.retry());
                 }
+            } finally {
+                cleanup();
             }
         });
 
         // If `RunnableFuture` was a `ListenableFuture` we could simply use `future.setFuture` to
         // propagate the result and cancellation, but instead add listener to propagate
         // cancellation to python thread, using the task executor which should invoke this in the
         // main thread (where this was originally called from)
-        future.addListener(new Runnable() {
-            @Override
-            public void run() {
-                if (future.isCancelled()) {
-                    Log.i(TAG, "Interrupting python thread");
-                    threadFuture.cancel(true);
-                }
+        future.addListener(() -> {
+            if (future.isCancelled()) {
+                Log.i(TAG, "Interrupting python thread");
+                threadFuture.cancel(true);
```

Comment:
Just some minor simplification here as I learned more about how these futures and runnables work.